connect/src/spirc.rs

```diff
@@ -680,10 +745,46 @@ impl SpircTask {
         self.state.get_position_ms() + diff as u32
     }
 
+    fn resolve_station(&self, uri: &str) -> Box<Future<Item = serde_json::Value, Error = MercuryError>> {
+        let radio_uri = format!("hm://radio-apollo/v3/stations/{}", uri);
+
+        self.resolve_uri(&radio_uri)
+    }
+
+    fn resolve_uri(&self, uri: &str) -> Box<Future<Item = serde_json::Value, Error = MercuryError>> {
+        let request = self.session.mercury().get(uri);
+
+        Box::new(request.and_then(move |response| {
+            let data = response.payload.first().expect("Empty payload on context uri");
+            let response: serde_json::Value = serde_json::from_slice(&data).unwrap();
+
+            Ok(response)
+        }))
+    }
+
+    fn update_tracks_from_context(&mut self) {
+        if let Some(ref context) = self.context {
+            self.context_fut = self.resolve_uri(&context.next_page_url);
+
+            let new_tracks = &context.tracks;
+            debug!("Adding {:?} tracks from context to playlist", new_tracks.len());
+            // Can we just push the new tracks and forget it?
+            let tracks = self.state.mut_track();
+            // tracks.append(new_tracks.to_owned());
+            for t in new_tracks {
+                tracks.push(t.to_owned());
```

Comment:
Hmmm, this could lead to a very large array in future if we get reconnection working and librespot is left running for a long time. It might be wise to set a limit of X many songs, an once the array hits that length, just shift `new_tracks.len()` from the `tracks` array to keep it at a known maximum size. The limit could be fairly high, say 200-500 uris, ie. more than enough for the end user not to notice, w/ regards to your comment @ashthespy...

Reply:
This is actually the `tracks` in the `Frame` - so the list that lets the client's view scroll into the new tracks (https://github.com/librespot-org/librespot/issues/57#issuecomment-451458008). Have to check and mimic the official clients behaviour - i.e how many tracks are there in the scrollback history if you leave a dailymix playing?

Reply:
will check, one sec

Reply:
10 previously played songs, but this limit has always irritated me, let's make it 20.

Reply:
Ah, you mean the current view, doubt we can change that. What we need to check is if the `Frame` holds all `pages` of the Context, or only the latest one. I.e if you leave a Dailymix running for a while, can you scroll all the way back to the first song?

Reply:
iirc you can't. will leave it playing for a couple of hours.

Reply:
confirmed, you can't (on Mac at least), can only go back 10 songs. Though I remember you mntioning on Android and windows you tested it up to 32, so go figure. We should just pick a limit and use that, I believe you suggested 42 tracks was suitable.

Reply:
it might also be worth investigating [std::collections::VecDeque](https://doc.rust-lang.org/std/collections/struct.VecDeque.html) as an alternative to storing the tracks in a `vec!`, as this would allow us to slice off `n` elements when loading new tracks, with `n` being `new_tracks.len()`

Reply:
This is going to take me a few more days to push - had to travel suddenly away from my dev machine. The route I have taken is to keep the last ten tracks and add the newly resolved ones. While it works, my implantation may not be the most efficient - the `Protobuf::RepeatedField` wrapper around `vec` doesn't expose `extend_from_slice` or similar - so need to push each track individually (unless I am being blind and have missed something?)

Reply:
I saw that, and investigated how protobuf is constructing the methods. Unfortunately it looks like we have to work within the constraints of the autogenerated bindings. I'd say the way you've done it if probably optimal given the constraints.Have tested and this is working well for me, I'd say merge it whenever.